### DISCUSSION TITLE
How do I DI IAnsiConsole?

### DISCUSSION THREAD

**Discussion by sandrock** [spectreconsole/spectre.console#1020] [Q&A]
I have a console app using dotnet 6 and dependency injection. 

I would like to declare the console as a service. That way the console is abstracted away and I can unit test it without garbage in the test console. Also I do not like the fact that all examples call static methods on `AnsiConsole`. I discovered the `IAnsiConsole` interface and thought it would help.

I tried this. So my other services can have it injected. 

```
services.AddOptions();
services.AddTransient<ConsoleApp>();              // my main thing
services.AddSingleton<IAnsiConsole, Recorder>();  // wrong way?
// and more...

var app = serviceProvider.GetService<ConsoleApp>();
app.Run(args).GetAwaiter().GetResult();
return app.ExitCode;
```

```
public class ConsoleApp : IConsoleApp
{
    public ConsoleApp(ILogger<ConsoleApp> log, IOptions<AppSettings> appSettings, ISomeClient client, IAnsiConsole console)
    {
        // ...
    }
}
```

It does not look like the `Recorder` class is the valid default implementation. I get this exception:

> Unhandled exception. System.InvalidOperationException: A circular dependency was detected for the service of type 'Spectre.Console.IAnsiConsole'.
Mynatime.CLI.ConsoleApp -> Spectre.Console.IAnsiConsole(Spectre.Console.Recorder) -> Spectre.Console.IAnsiConsole

How are we supposed to proceed then?

I looked at [How to integrate with .NET Generic Host?](https://github.com/spectreconsole/spectre.console/discussions/380) but I don't understand anything there. Seems related to Spectre.Cli, not Spectre.Console. 
I looked at [Is there an alternative to the global/static AnsiConsole for testing?](https://github.com/spectreconsole/spectre.console/discussions/571). Unfortunately, it only answers the testing side, not the app side.

**Comment by sandrock** [comment]
This seem to work for simple `WriteLine`s.

```
services.AddSingleton<IAnsiConsole>((x) => AnsiConsole.Create(new AnsiConsoleSettings()
{
    Ansi = AnsiSupport.Detect,
    ColorSystem = ColorSystemSupport.Detect,
    Out = (IAnsiConsoleOutput) new AnsiConsoleOutput(System.Console.Out)
}));
```

Not sure it's the way it was planned for.

**Comment by phil-scott-78** [comment]
Have you checked out the [docs for DI](https://spectreconsole.net/cli/commandapp)? Spectre instantiates the commands, not the container, so you need to tell it about your container.

It'll also properly wire up `IAnsiConsole` too

**Reply by sandrock** [reply]
Hello. Thank you for your time.

I did look at that but I am not using `Spectre.Console.Cli`. 

> It'll also properly wire up IAnsiConsole too

Perhaps I should look into this code to see how is works internally.

**Comment by patriksvensson** [comment]
You can inject the console like this: `services.AddSingleton<IAnsiConsole>(AnsiConsole.Console)`
For testing, I recommend the `TestConsole` in the `Spectre.Console.Testing` package.
